File: bridges/python/src/sdk/tool_manager.py
from typing import Optional, Type

from .base_tool import BaseTool
from .mira import mira
from .utils import format_file_path


class MissingToolSettingsError(Exception):
    def __init__(self, missing: list[str], settings_path: str):
        super().__init__(f"Missing tool settings: {', '.join(missing)}")
        self.missing = missing
        self.settings_path = settings_path


class ToolManager:
    @staticmethod
    def init_tool(tool_class: Type[BaseTool]) -> BaseTool:
        tool = tool_class()
        missing = tool.get_missing_settings()

        if missing:
            mira.answer(
                {
                    "key": "bridges.tools.missing_settings",
                    "data": {
                        "tool_name": tool.alias_tool_name,
                        "missing": ", ".join(missing.get("missing", [])),
                        "settings_path": format_file_path(
                            missing.get("settings_path", "")
                        ),
                    },
                    "core": {
                        "should_stop_skill": True,
                    },
                }
            )
            raise MissingToolSettingsError(
                missing.get("missing", []),
                missing.get("settings_path", ""),
            )

        return tool


def is_missing_tool_settings_error(error: Exception) -> bool:
    return isinstance(error, MissingToolSettingsError)

File: bridges/nodejs/src/sdk/tool-manager.ts
import { formatFilePath } from '@sdk/utils'
import { Tool } from '@sdk/base-tool'
import { reportToolOutput } from '@sdk/tool-reporter'

export class MissingToolSettingsError extends Error {
  missing: string[]
  settingsPath: string

  constructor(missing: string[], settingsPath: string) {
    super(`Missing tool settings: ${missing.join(', ')}`)
    this.name = 'MissingToolSettingsError'
    this.missing = missing
    this.settingsPath = settingsPath
  }
}

export const isMissingToolSettingsError = (
  error: unknown
): error is MissingToolSettingsError => {
  return error instanceof MissingToolSettingsError
}

export default class ToolManager {
  static async initTool<TTool extends Tool>(
    ToolClass: new () => TTool
  ): Promise<TTool> {
    const tool = new ToolClass()
    const missing = tool.getMissingSettings()

    if (missing) {
      try {
        await reportToolOutput({
          key: 'bridges.tools.missing_settings',
          data: {
            tool_name: tool.aliasToolName,
            missing: missing.missing.join(', '),
            settings_path: formatFilePath(missing.settingsPath)
          },
          core: {
            should_stop_skill: true
          }
        })
      } catch (error) {
        console.warn(
          `[MIRA_TOOL_LOG] Failed to report missing tool settings: ${
            (error as Error).message
          }`
        )
      }
      throw new MissingToolSettingsError(missing.missing, missing.settingsPath)
    }

    return tool
  }
}
